### PR TITLE
Login

### PR DESCRIPTION
**Pull request by tomaszmrugalski**
This is now ready for review. This PR adds an ability to log in. This using flask-login. This will allow us to restrict access to certain parts of the webpage for registered users, possibly with certain role. There are currently 4 roles: user, station-owner, admin and banned.

### CODE REVIEW

File: server/app/controllers/login.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        return self.auth
+
+    def is_anonymous(self):
+        return not self.auth
+
+    def get_id(self):
+        return self.id
+
+
+@app.route("/login", methods=["GET", "POST"])
```
Comment (fivitti) [inline server/app/controllers/login.py:52]: Login contains password. It should be send only using POST method.

File: server/app/controllers/login.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            app.logger.info("Login failed: attempt to login into disabled account %s" % form.username.data)
+            flash("Account disabled.")
+            return redirect(url_for("login"))
+
+        app.logger.info("Login successful for user %s" % form.username.data)
+        login_user(u, remember = form.remember.data)
+
+        next_page = request.args.get("next")
+        if not next_page or url_parse(next_page).netloc != "":
+            next_page = url_for("login")
```
Comment (fivitti) [inline server/app/controllers/login.py]: Maybe "index" page will be better to redirect after login?

File: server/app/controllers/login.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        return redirect(next_page)
+
+    return render_template("login.html", form=form)
+
+login = LoginManager(app)
```
Comment (fivitti) [inline server/app/controllers/login.py]: Is it good? You overwrite login method reference.

File: server/app/repository.py [hunk 4]
```diff
@@ -379,6 +393,33 @@ def read_station_secret(self, station_id: StationId) -> Optional[bytes]:
...
         return bytes(row["secret"])
 
+    def user_role_to_enum(self, role: str) -> UserRole:
+        """Converts string to UserRole enum"""
+        for u in UserRole:
+            if u.name == role.upper():
+                return u
+
+        # If we were not able to figure out, let's assume it's a regular user
+        return UserRole.REGULAR
```
Comment (fivitti) [inline server/app/repository.py]: I think that you should throw here ```LookupError```. It isn't standard situation that role doesn't exist and shouldn't occur in normal flow.If you suppress it then it may be complicate further development.

File: server/app/repository.py [hunk 5]
```diff
@@ -379,6 +393,33 @@ def read_station_secret(self, station_id: StationId) -> Optional[bytes]:
...
+
+    @use_cursor
+    def read_user(self, id: int = 0, username: str = "") -> Optional[User]:
```
Comment (fivitti) [inline server/app/repository.py]: It isn't pytonic. You should use single argument type (as in read_station).
```typing``` module has special decorator for it: [@overload](https://docs.python.org/3/library/typing.html#typing.overload)

File: server/app/repository.py [hunk 6]
```diff
@@ -379,6 +393,33 @@ def read_station_secret(self, station_id: StationId) -> Optional[bytes]:
...
+            cursor.execute(query, (id,))
+        elif username:
+            query = "SELECT id, username, digest, email, role FROM users WHERE username = %s"
+            cursor = self._cursor
+            cursor.execute(query, (username,))
+
+        row = cursor.fetchone()
+        u = None
+        if row:
+            row['role'] = self.user_role_to_enum(row['role'])
```
Comment (fivitti) [inline server/app/repository.py:420]: What is returned from database? How it is interpreted by psycopg2? Probably it returns string and will be better to keep it in this form. Instead of using subclass of ```Enum``` you may to use ```Literal``` from ```typing``` module.

File: server/satnogs-web.py [hunk 7]
```diff
@@ -1,5 +1,7 @@
 #!/usr/bin/env python3
 from app import app
 
+app.config['SECRET_KEY'] = 'the earth is flat'
+
 if __name__ == '__main__':
-    app.run(debug=True, port=5000)
+    app.run(debug=True, port=8080)
```
Comment (fivitti) [inline server/satnogs-web.py:7]: Uhh, why? You make me to reconfigure my IDE.

File: server/satnogs-web.py [hunk 8]
```diff
@@ -1,5 +1,7 @@
 #!/usr/bin/env python3
 from app import app
 
+app.config['SECRET_KEY'] = 'the earth is flat'
```
Comment (fivitti) [inline server/satnogs-web.py:4]: Where it is used? Is this value secure? Should we commit it?

File: doc/users.md [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+Users information has been introduced in [schema update 07](../server/db/satnogs-07.psql). There's a new table that holds
+user-id, username, salted SHA256 digest, and user role. Until we add flask-admin (or some other management panel), the
+process of adding new users is manual.
+
+To add a new user:
+
+1. log into the server using ssh
+2. psql satnogs
+3. add entry for new user
```
Comment (fivitti) [inline doc/users.md:13]: One user should be created by default.

Comment (fivitti) [inline doc/users.md:13]: Has user the possibility to change password?

File: server/app/controllers/login.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+from app.repository import Repository, UserRole
+
+class LoginForm(FlaskForm):
+    username = StringField("Your login name", validators=[InputRequired(), Length(min=3, max=20)])
+    password = PasswordField("Your password", validators=[InputRequired(), Length(min=5, max=80)])
+    remember = BooleanField("Remember me")
+
+    submit = SubmitField("Sign In")
+
+class SatnogsUser(UserMixin):
```
Comment (fivitti) [inline server/app/controllers/login.py]: We plan to abandon "Satnogs" name for out project. Maybe "ApplicationUser" will be better?

File: server/app/controllers/login.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+    # The following fields are useful for Flask-login
+    # It's basically a reimplementation of UserMixin
+
+    def __init__(self, user):
+        # I'm sure there's easier way to do it. There's some magic involving **...
```
Comment (fivitti) [inline server/app/controllers/login.py]:
```python
for key, value in user.items():
    setattr(self, key, value)
```
But maybe ```user``` should be store as member of this class?

File: server/app/repository.py [hunk 12]
```diff
@@ -379,6 +393,33 @@ def read_station_secret(self, station_id: StationId) -> Optional[bytes]:
             return None
         return bytes(row["secret"])
 
+    def user_role_to_enum(self, role: str) -> UserRole:
+        """Converts string to UserRole enum"""
+        for u in UserRole:
```
Comment (fivitti) [inline server/app/repository.py]: In ```utils.py``` you have ```first``` function.

```python
return first(lambda u: u.name == role.upper(), UserRole)
```
but it may return None then you need test for it.

File: doc/users.md [hunk 13]
```diff
@@ -0,0 +1,29 @@
+[doc](../README.md) > Users
+
+# Users Management
+
+Users information has been introduced in [schema update 07](../server/db/satnogs-07.psql). There's a new table that holds
```
Comment (fivitti) [inline doc/users.md]: It is ```satnogs-08.psql``` file.

**Review by fivitti — CHANGES_REQUESTED** [review]

File: server/app/controllers/login.py (hunk 1 above)
Comment (tomaszmrugalski) [reply server/app/controllers/login.py:52]: When user navigates to the page, it's always by GET. The form contents is being sent using POST. The whole transmission is protected with TLS.

I tried to update the code as you requested (accept post only, but then it was no longer possible to log in.

File: server/app/controllers/login.py (hunk 2 above)
Comment (tomaszmrugalski) [reply server/app/controllers/login.py]: Staying on login page has the benefit of showing the user details as a confirmation that the login was indeed successful. But I don't care much about that, so can update as proposed.

This required a bit gymnastics in the tests, but I've figured it out.

File: server/app/controllers/login.py (hunk 3 above)
Comment (tomaszmrugalski) [reply server/app/controllers/login.py]: Renamed.

File: server/app/repository.py (hunk 4 above)
Comment (tomaszmrugalski) [reply server/app/repository.py]: The code now throws as suggested. Also added unittest for this.

File: server/app/repository.py (hunk 5 above)
Comment (tomaszmrugalski) [reply server/app/repository.py]: I'm willing to update the code, but I don't understand how it's supposed to look. You mentioned read_station. It takes one parameter: id, which is of StationId type. There's no overloading there.

File: server/satnogs-web.py (hunk 7 above)
Comment (tomaszmrugalski) [reply server/satnogs-web.py:7]: I have other projects that are being run locally. All of them use 8080. But if it's a problem, I can withdraw this change. 8080 is a natural port for web traffic. (80 being a privileged port requires root access, so 8080 is the second best).

File: server/satnogs-web.py (hunk 8 above)
Comment (tomaszmrugalski) [reply server/satnogs-web.py:4]: It's necessary. Otherwise there's exception when flask-login attempts to use session to store successful login. Added text explaining that:

If you're curious, here's the full stack:
```python
venv/lib/python3.7/site-packages/werkzeug/test.py:1016: in post
    return self.open(*args, **kw)
venv/lib/python3.7/site-packages/flask/testing.py:227: in open
    follow_redirects=follow_redirects,
venv/lib/python3.7/site-packages/werkzeug/test.py:970: in open
    response = self.run_wsgi_app(environ.copy(), buffered=buffered)
venv/lib/python3.7/site-packages/werkzeug/test.py:861: in run_wsgi_app
    rv = run_wsgi_app(self.application, environ, buffered=buffered)
venv/lib/python3.7/site-packages/werkzeug/test.py:1096: in run_wsgi_app
    app_rv = app(environ, start_response)
venv/lib/python3.7/site-packages/flask/app.py:2463: in __call__
    return self.wsgi_app(environ, start_response)
venv/lib/python3.7/site-packages/flask/app.py:2449: in wsgi_app
    response = self.handle_exception(e)
venv/lib/python3.7/site-packages/flask/app.py:1866: in handle_exception
    reraise(exc_type, exc_value, tb)
venv/lib/python3.7/site-packages/flask/_compat.py:39: in reraise
    raise value
venv/lib/python3.7/site-packages/flask/app.py:2446: in wsgi_app
    response = self.full_dispatch_request()
venv/lib/python3.7/site-packages/flask/app.py:1951: in full_dispatch_request
    rv = self.handle_user_exception(e)
venv/lib/python3.7/site-packages/flask/app.py:1820: in handle_user_exception
    reraise(exc_type, exc_value, tb)
venv/lib/python3.7/site-packages/flask/_compat.py:39: in reraise
    raise value
venv/lib/python3.7/site-packages/flask/app.py:1949: in full_dispatch_request
    rv = self.dispatch_request()
venv/lib/python3.7/site-packages/flask/app.py:1935: in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
app/controllers/login.py:72: in login
    flash("Invalid username.")
venv/lib/python3.7/site-packages/flask/helpers.py:423: in flash
    session["_flashes"] = flashes
venv/lib/python3.7/site-packages/werkzeug/local.py:350: in __setitem__
    self._get_current_object()[key] = value
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <NullSession {}>, args = ('_flashes', [('message', 'Invalid username.')]), kwargs = {}

    def _fail(self, *args, **kwargs):
        raise RuntimeError(
>           "The session is unavailable because no secret "
            "key was set.  Set the secret_key on the "
            "application to something unique and secret."
        )
E       RuntimeError: The session is unavailable because no secret key was set.  Set the secret_key on the application to something unique and secret.

venv/lib/python3.7/site-packages/flask/sessions.py:103: RuntimeError

```

File: doc/users.md (hunk 9 above)
Comment (tomaszmrugalski) [reply doc/users.md:13]: Default users, especially with passwords set is a security nightmare. See [this question](https://security.stackexchange.com/questions/141157/is-it-dangerous-to-use-default-router-admin-passwords-if-only-trusted-users-are). If you insist on doing that, I can add a user, but it will not have password set by default.

In full blown systems there's usually a mechanism that forces people to change their password on first login on provided passwords. But I don't want to waste time to implement such things here.

Comment (tomaszmrugalski) [reply doc/users.md:13]: Sadly, not yet. I was thinking that since we have only 2 users, this is sufficient for now. I can imagine a profile editing panel, where user can edit his first, last name (we don't have that yet), edit his email, change password and possibly do other things. All of that IMHO is a distraction right now. Implementing such features right now does not add value to the project and is a waste of time right now.

This should be pushed to a separate ticket that sadly will be ignored for a while.

File: server/app/controllers/login.py (hunk 10 above)
Comment (tomaszmrugalski) [reply server/app/controllers/login.py]: That name is a too long IMHO, but I don't care, so renamed as suggested.

File: server/app/controllers/login.py (hunk 11 above)
Comment (tomaszmrugalski) [reply server/app/controllers/login.py]: But that code would only copy existing keys. It would ensure that all five fields are set. Anyway, updated as suggested.

File: server/app/repository.py (hunk 12 above)
Comment (tomaszmrugalski) [reply server/app/repository.py]: Updated as suggested. Also testing if None is returned.

On a related note, it's super confusing that we have several utils.py. We should rename rename one of them one day. One way to deconflict the names would be to rename server/tests/utils.py to server/tests/test_utils.py.

File: doc/users.md (hunk 13 above)
Comment (tomaszmrugalski) [reply doc/users.md]: Good catch! It was 07 when I was developing it, but your code (with your -07) got to master first.

File: server/app/controllers/login.py (hunk 1 above)
Comment (fivitti) [reply server/app/controllers/login.py:52]: Ok, then your function has two responsibilities: handle "login" page and validate and login.
One responsibility use GET method, second POST. 

In Flask you may split it and use multiple routes.
```python
@app.route("/login", methods=["GET"])
def login():
   if current_user.is_authenticated:
        return render_template("login.html", user = current_user)
    form = LoginForm()
    return render_template("login.html", form=form)

@app.route("/login", methods=["POST"])
def login_validate():
   '''Validate and login''''
```
But in this case we may keep it as is.

File: doc/users.md (hunk 9 above)
Comment (tomaszmrugalski) [reply doc/users.md:13]: Created #69 to not forget about this.

File: server/app/repository.py (hunk 5 above)
Comment (fivitti) [reply server/app/repository.py]: First proposal:
```python
def read_user(self, user:  Union[int, str]) -> Optional[User]:
   if type(user) == str:
      ...
   else:
       ...
```
you may also add typing:
```python
@overload
def read_user(self, user: int) -> Optional[User]:
    pass
@overload
def read_user(self, user: str) -> Optional[User]:
    pass
def read_user(self, user: Union[str, int]) -> Optional[User]:
   # implementation
```

File: server/app/repository.py [hunk 14]
```diff
@@ -75,6 +77,19 @@ class StationPhoto(TypedDict):
     filename: str
     descr: str
 
+class UserRole(Enum):
```
Comment (fivitti) [inline server/app/repository.py:80]: Check for possibility use string or Literal (better) for it. Maybe you avoid conversion from int to string and from string to int when fetch data from DB. In dynamic typed languages as Python Enums have less advantages then static typed languages.

**Review by fivitti — CHANGES_REQUESTED** [review]

File: server/app/repository.py (hunk 12 above)
Comment (fivitti) [reply server/app/repository.py]: Ok, good suggestion.

Comment (tomaszmrugalski) [reply server/app/repository.py]: Renamed.

File: server/app/repository.py (hunk 14 above)
Comment (tomaszmrugalski) [reply server/app/repository.py:80]: For some reason there are two aspects here in this thread.

The first one is about read_user. I've updated the code to use single union parameter.

Comment (tomaszmrugalski) [reply server/app/repository.py:80]: Regarding, the second aspect (replace Enum with Literals), here's my answer:

There is no conversion from integers (at least not in our code. Perhaps somewhere deep in psycopg2, but I don't care). The underlying enum type in postgres is returned as a string and this is what cursor.fetchone() returns. I can get rid of the Enum and replace it with a class that would have 4 literals defined, but I don't understand why it would be any better. I can iterate over Enum to get all values. This may be useful in the future to display some help information, sanity check admin panel data etc.

Can you point me to an example class somewhere that uses Literals as enum values? I don't understand what exaclty you propose here. I want something that would accept only explicit 4 values and would throw for everything else. I find it useful if it's possible to programmatically get a list of all possible values. Having explicit expression in the code everywhere values are used (e.g. you grep for UserRole.ADMIN and you find ALL references) is another big bonus.

Comment (tomaszmrugalski) [reply server/app/repository.py:80]: I've tried to experiment with Literal, but after a while I strongly reject the idea as strongly inferior to Enum. Here are my reasons:

1. You can't iterate over Literal. There is a good reason for that. Soon we'll be implementing an admin panel and we'll need a list of all possible roles.
2. I still haven't heard any convincing argument why to use Literal at all.
3. I did an experiment and wanted to try Literal in practice. I've modified the UserRole(Enum) with UserRole = Literal[...] and got this exception. I must admit it's confusing as hell for a simple enum.

```
app/repository.py:397: in user_role_to_enum
    u = first(lambda u: u.name == role.upper(), UserRole)
app/utils.py:28: in first
    for item in items:
/usr/lib/python3.7/typing.py:254: in inner
    return func(*args, **kwds)
/usr/lib/python3.7/typing.py:629: in _getitem_
    params = tuple(_type_check(p, msg) for p in params)
/usr/lib/python3.7/typing.py:629: in <genexpr>
    params = tuple(_type_check(p, msg) for p in params)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

arg = 0, msg = 'Parameters to generic types must be types.', is_argument = True

    def _type_check(arg, msg, is_argument=True):
        """Check that the argument is a type, and return it (internal helper).
    
        As a special case, accept None and return type(None) instead. Also wrap strings
        into ForwardRef instances. Consider several corner cases, for example plain
        special forms like Union are not valid, while Union[int, str] is OK, etc.
        The msg argument is a human-readable error message, e.g::
    
            "Union[arg, ...]: arg should be a type."
    
        We append the repr() of the actual value (truncated to 100 chars).
        """
        invalid_generic_forms = (Generic, _Protocol)
        if is_argument:
            invalid_generic_forms = invalid_generic_forms + (ClassVar, )
    
        if arg is None:
            return type(None)
        if isinstance(arg, str):
            return ForwardRef(arg)
        if (isinstance(arg, _GenericAlias) and
                arg.__origin__ in invalid_generic_forms):
            raise TypeError(f"{arg} is not valid as type argument")
        if (isinstance(arg, _SpecialForm) and arg not in (Any, NoReturn) or
                arg in (Generic, _Protocol)):
            raise TypeError(f"Plain {arg} is not valid as type argument")
        if isinstance(arg, (type, TypeVar, ForwardRef)):
            return arg
        if not callable(arg):

              raise TypeError(f"{msg} Got {arg!r:.100}.")

E           TypeError: Parameters to generic types must be types. Got 0.

/usr/lib/python3.7/typing.py:142: TypeError
```

4. Continuing with my experiment, I've removed user_role_to_enum and all references to it, and replaced UserRole.ADMIN with 'admin' literal. Now the tests started to fail:
```
E       AssertionError: 'admin' != 'ADMIN'
E       - admin
E       + ADMIN
``` 
Which is a proof that in the process we also lost value checking in the process. Yes, this could be solved by using capital 'ADMIN' everywhere, but this puts the burden of remembering if it's capital or lowercase on user. I was using Visual Studio Code. The env did not stop me and didn't give any hints.

5. Literal is a recent addition. It's in typing in python 3.8, but in typing_extensions in 3.7. Don't know if it's available in 3.6.

As such, I discarded my experimental changes and strongly feel that going forward with Enum is the way to go.

File: server/app/repository.py (hunk 5 above)
Comment (tomaszmrugalski) [reply server/app/repository.py]: Updated as suggested. There's one Union parameter.

**Review by fivitti — APPROVED** [review]
